CI: Cache dependencies in a container image

nicoonoclaste opened the pull request.

nicoonoclaste committed:
CI: Add templated Dockerfiles for Linux environments

We can install all test dependencies (incl. pygame's build dependencies where
necessary) in a container image, and cache those across builds.

See https://medium.com/cirruslabs/introducing-container-builder-for-cirrus-ci-80b9234f007

nicoonoclaste committed:
CI: Generate Dockerfiles

nicoonoclaste committed:
CI: Use cacheable container images for Linux environments

nicoonoclaste committed:
CI: Check that the CI Dockerfiles are up-to-date

nicoonoclaste committed:
CI/Dockerfile: Use the correct Python interpreter name in templates

This is nicer than detecting it at runtime, and can work with Windows

nicoonoclaste committed:
CI/Dockerfile: Detect if pygame's build deps are necessary

Also, get rid of the separate template.Dockerfile file

nicoonoclaste committed:
CI/Dockerfiles: Generate Windows images

nicoonoclaste committed:
CI: Commit Dockerfiles for Windows

nicoonoclaste committed:
CI: Use cached, custom images for Windows tests

nicoonoclaste committed:
CI: Rename Dockerfiles (: isn't valid in Windows filenames)

nbraud commented:
```
At line:1 char:145
+ ... thon.exe -m pip install --upgrade-strategy eager -U pytest && C:Pytho ...
+                                                                ~~
The token '&&' is not a valid statement separator in this version.
```
How dare!

nicoonoclaste committed:
CI/Dockerfile: Sequence commands differently on Linux and Windows

For some reason, Docker commands run in Powershell, and Powershell doesn't
implement `&&`...

nicoonoclaste committed:
CI/Dockerfile: How does one even delete folders in Powershell?

nicoonoclaste committed:
CI/Dockerfile: Split utility functions to a separate file

nicoonoclaste committed:
CI/Dockerfile: Get rid of the separate install-debian-deps script

nicoonoclaste committed:
CI: Pull test dependencies from requirements-tests.txt

nicoonoclaste committed:
CI: Regenerate Dockerfiles

nbraud commented:
Retargeting against the main PPB repository.
